[FSDP] Add no_sync() context manager

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72578 [ZeRO] Add ctor support for multiple param groups
* **#72446 [FSDP] Add no_sync() context manager**

**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

**Test Plan**
- `_test_no_sync()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun4 python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

Differential Revision: [D34085750](https://our.internmc.facebook.com/intern/diff/D34085750)

Commit:
[FSDP] Add no_sync() context manager

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
I generalized `_test_grad_acc()` from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)) to parameterize over the number of iterations for which to accumulate gradients instead of being fixed to exactly 2 (1 without sync and 1 with sync).

I have validated it for world sizes of 2 and 4:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
```

**NOTE:** I had to use double precision in order to avoid errors from floating point drift. With single precision floats, the tests fail due to slight discrepancies.

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -0,0 +1,146 @@
+            "Accumulate for at least 2 iterations to be nontrivial"
+        self._test_no_sync(batch_dim=1, num_iters_to_acc=num_iters_to_acc)
+
+    def _test_communication(self):
```
We can test the communication costs with benchmarks, is this test necesssary / does it add value in terms of testing the API?

awgu replied:
This test is pretty interesting to me. When you wrote this comment, I had not yet pushed what I had, so it probably did not make much sense.

For example, through this test, I was able to see that PyTorch `FSDP` makes an additional all-gather call in the forward pass compared to Fairscale `FSDP`. Also, only thinking about the specification of `no_sync()`, I think testing that it indeed saves on communication primitives makes sense.

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -0,0 +1,146 @@
+    sys.exit(0)
+
+
+class TestGradAcc(FSDPTest):
```
Can we test with CPU offload, backward prefetch, etc as well just to make sure things are working as expected?

awgu replied:
Testing with those orthogonal features makes sense to me! (though I may need some time to familiarize myself)

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -0,0 +1,146 @@
+                loss.backward()
+                losses.append(loss)
+        output = fsdp_model(*batches[-1])
+        loss = fsdp_model.module.get_loss(batches[-1], output)
```
Should we run an optimizer.step() after no_sync() is exited just to make sure things work fine?

awgu replied:
Added.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
+            yield
+        finally:
+            for m, old_flag in old_flags:
+                assert not m._require_backward_grad_sync
```
Add an error msg?

awgu replied:
Done.

rohan-varma reviewed (commented):
Oops didn't see this was a draft and I already wrote some comments, so just sending them your way for consideration!

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
I generalized `_test_grad_acc()` from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)) to parameterize over the number of iterations for which to accumulate gradients instead of being fixed to exactly 2 (1 without sync and 1 with sync).

I tested it for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
```

**To-Do**
- I had to use double precision in order to avoid errors from floating point drift. With single precision floats, the tests fail due to slight discrepancies. Is this a problem?
- Fairscale tests the communication cost of `no_sync()` by counting the invocations of the collective communication primitives (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)). An equivalent test should be included here.
- Do I include the block comment describing the test in the test wrapper or the test helper?

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -810,7 +813,12 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
 
-        self._free_full_params([param])
+        if self._require_backward_grad_sync:
+            # We do not free the full parameters in a `no_sync()` context since
```
I realized that this was legacy for the Fairscale implementation and was removed in more recent commits. I have removed this conditional.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
+        old_flags = []
+        for m in self.modules():
+            if isinstance(m, FullyShardedDataParallel):
+                old_flags.append((m, m._require_backward_grad_sync))
```
This is minor, but in my understanding, Python generator ordering is not guaranteed. In practice, `torch.nn.Module.modules()` probably does always return the same ordering, but this might be relying on non-specified behavior.

Is there any downside to maintaining an extra reference to each module? The modules should not be being deleted anyway (so these references should not be the critical ones preventing garbage collection)?

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
+            accumulate the full model gradients (instead of gradient shards)
+            until the eventual sync.
+
+        .. note:: Gradient accumulation can be done without this context,
```
Good catch.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
+    @contextmanager
+    def no_sync(self) -> Generator:
+        """
+        A context manager to disable gradient synchronizations across FSDP
```
Done.

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
I generalized `_test_grad_acc()` from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)) to parameterize over the number of iterations for which to accumulate gradients instead of being fixed to exactly 2 (1 without sync and 1 with sync).

I tested it for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
```

**To-Do**
- I had to use double precision in order to avoid errors from floating point drift. With single precision floats, the tests fail due to slight discrepancies. Is this a problem?
- Add optimizer step to the main parity test.
- Determine why PyTorch `FSDP` makes an additional forward-pass all-gather compared to Fairscale `FSDP`.
- Add tests for checking `no_sync()` compatibility with CPU offloading and backward prefetch.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
I generalized `_test_grad_acc()` from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)) to parameterize over the number of iterations for which to accumulate gradients instead of being fixed to exactly 2 (1 without sync and 1 with sync).

I tested it for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
```

**To-Do**
- I had to use double precision in order to avoid errors from floating point drift. With single precision floats, the tests fail due to slight discrepancies. Is this a problem?
- Add optimizer step to the main parity test.
- Determine why PyTorch `FSDP` makes an additional forward-pass all-gather compared to Fairscale `FSDP`.
- Add tests for checking `no_sync()` compatibility with CPU offloading and backward prefetch.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
I generalized `_test_grad_acc()` from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)) to parameterize over the number of iterations for which to accumulate gradients instead of being fixed to exactly 2 (1 without sync and 1 with sync).

I tested it for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
```

**To-Do**
- I had to use double precision in order to avoid errors from floating point drift. With single precision floats, the tests fail due to slight discrepancies. Is this a problem?
- Add optimizer step to the main parity test.
- Determine why PyTorch `FSDP` makes an additional forward-pass all-gather compared to Fairscale `FSDP`.
- Add tests for checking `no_sync()` compatibility with CPU offloading and backward prefetch.

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -9,6 +10,7 @@
+    Generator,
     List,
     Optional,
     Generator,
```
lint error is real

awgu replied:
Good catch. My VSCode Python extension broke starting yesterday, and I was finally able to fix it by reverting to an older version. :sweat_smile:

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
I generalized `_test_grad_acc()` from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)) to parameterize over the number of iterations for which to accumulate gradients instead of being fixed to exactly 2 (1 without sync and 1 with sync).

I tested it for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
```

**To-Do**
- I had to use double precision in order to avoid errors from floating point drift. With single precision floats, the tests fail due to slight discrepancies. Is this a problem?
- Add optimizer step to the main parity test.
- Determine why PyTorch `FSDP` makes an additional forward-pass all-gather compared to Fairscale `FSDP`.
- Add tests for checking `no_sync()` compatibility with CPU offloading and backward prefetch.

[ghstack-poisoned]

mrshenli reviewed (commented):
A meta question:

`no_sync` is certainly useful feature for users, but IIUC it is not very friendly to FSDP, as we will need to either run H2D/D2H and allgather to collect parameters for every micro-batch (costly) or keep params + grads in CUDA memory (won't work for large models). 

I wonder if there are plans to add real micro-batching for FSDP? I.e., FSDP still takes one mini-batch, and then each FSDP unit chunk the input into multiple micro-batches and process all of them one-by-one before moving on to the next FSDP unit. More specifically, 

current `no_sync` does: FSDP_unit1(micro_batch1) -> FSDP_unit2(micro_batch1) -> FSDP_unit1(micro_batch2) -> FSDP_unit2(micro_batch2)

real micro-batching does : FSDP_unit1(micro_batch1) -> FSDP_unit1(micro_batch2) -> FSDP_unit2(micro_batch1) -> FSDP_unit2(micro_batch2)

The benefit of real micro-batching is that we only need to run one D2H/H2D + allgather across all micro-batches for each FSDP unit's parameters, rather than running D2H/H2D + allgather once for every micro-batch.

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
I generalized `_test_grad_acc()` from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)) to parameterize over the number of iterations for which to accumulate gradients instead of being fixed to exactly 2 (1 without sync and 1 with sync).

I tested it for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
```

**To-Do**
- I had to use double precision in order to avoid errors from floating point drift. With single precision floats, the tests fail due to slight discrepancies. Is this a problem?
- Add optimizer step to the main parity test.
- Determine why PyTorch `FSDP` makes an additional forward-pass all-gather compared to Fairscale `FSDP`.
- Add tests for checking `no_sync()` compatibility with CPU offloading and backward prefetch.

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -0,0 +1,239 @@
+            torch.testing.assert_allclose(ref_grad, acc_grad)
+
+        # Restore the default dtype
+        torch.set_default_dtype(old_default_dtype)
```
Might want to do a try: finally block to ensure this gets reset appropriately even if the test somehow fails

awgu replied:
Great point!

rohan-varma commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -0,0 +1,239 @@
+            if isinstance(m, FSDP) and len(m.params) > 0:
+                num_fsdp += 1
+
+        # Not all PyTorch versions have the `_base` methods available
```
I'm not sure if we need this as we don't test against all PyTorch versions, only PT master which should have both of these. In PT FSDP codebase I believe we unconditionally use them too

awgu replied:
Yup, fixed.

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
- `_test_grad_acc()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_no_sync_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
```

**To-Do**
- I had to use double precision in order to avoid errors from floating point drift in `_test_grad_acc()`. With single precision floats, the tests fail due to slight discrepancies. Is this a problem? I save and restore the old default `dtype`, so hopefully the `dtype` change is actually contained to the test.
- Determine why PyTorch `FSDP` makes an additional forward-pass all-gather compared to Fairscale `FSDP`.
- Add optimizer step to the main parity test.
- Add tests for checking `no_sync()` compatibility with CPU offloading and backward prefetch.
- Remember to run existing FSDP tests to check for regression.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.


**Test Plan**
- `_test_grad_acc()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

**To-Do**
- Investigate if non-determinism can substitute for using double precision to avoid floating point drift in `_test_grad_acc()`.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

See this issue: https://github.com/pytorch/pytorch/issues/72541

**Test Plan**
- `_test_grad_acc()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

**To-Do**
- Investigate if non-determinism can substitute for using double precision to avoid floating point drift in `_test_grad_acc()`.

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

See this issue: https://github.com/pytorch/pytorch/issues/72541

**Test Plan**
- `_test_grad_acc()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

**Notes**
I investigated the floating-point drift in `_test_grad_acc()`. Using `torch.use_deterministic_algorithms(True)` does not help the issue, which makes sense since none of the non-deterministic algorithms (see [here](https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)) should be used in the transformer model. The current solution is to use double precision just for the `_test_grad_acc()` test.

Differential Revision: [D34085750](https://our.internmc.facebook.com/intern/diff/D34085750)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

See this issue: https://github.com/pytorch/pytorch/issues/72541

**Test Plan**
- `_test_grad_acc()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

**Notes**
I investigated the floating-point drift in `_test_grad_acc()`. Using `torch.use_deterministic_algorithms(True)` does not help the issue, which makes sense since none of the non-deterministic algorithms (see [here](https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)) should be used in the transformer model. The current solution is to use double precision just for the `_test_grad_acc()` test.

Differential Revision: [D34085750](https://our.internmc.facebook.com/intern/diff/D34085750)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

awgu commented on test/distributed/fsdp/test_fsdp_grad_acc.py:
```diff
@@ -0,0 +1,167 @@
+        assert num_iters_to_acc >= 2, \
+            "Accumulate for at least 2 iterations to be nontrivial"
+        self._test_no_sync(
+            batch_dim=1,
```
The only purpose of `test_no_sync()` is to manually pass in `batch_dim=1` to `_test_no_sync()`. If this seems unnecessary, I can coalesce the two methods.

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

See this issue: https://github.com/pytorch/pytorch/issues/72541

**Test Plan**
- `_test_no_sync()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun4 python test/distributed/fsdp/test_fsdp_grad_acc.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

**Notes**
I investigated the floating-point drift in `_test_no_sync()`. Using `torch.use_deterministic_algorithms(True)` does not help the issue, which makes sense since none of the non-deterministic algorithms (see [here](https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)) should be used in the transformer model. The current solution is to use double precision just for the `_test_no_sync()` test.

Differential Revision: [D34085750](https://our.internmc.facebook.com/intern/diff/D34085750)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

awgu commented:
I changed the test filename from `test_fsdp_grad_acc.py` to `test_fsdp_no_sync.py` since the only way to do gradient accumulation at this point is to use `no_sync()`. Fairscale seems to have another way without `no_sync()` (see https://github.com/pytorch/pytorch/issues/72185). We can rename the file and test class if/when we upstream that.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1175,6 +1181,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
             raise ValueError(msg)
 
+    @contextmanager
+    def no_sync(self) -> Generator:
```
Can we also file an issue to refactor/deduplicate the no_sync in DDP and FSDP? They look pretty similar and we can probably provide some sort of config / lambda to take care of DDP and FSDP implementation specific details.

awgu replied:
https://github.com/pytorch/pytorch/issues/72589

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

**Test Plan**
- `_test_no_sync()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun4 python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

Differential Revision: [D34085750](https://our.internmc.facebook.com/intern/diff/D34085750)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

zhaojuanmao reviewed (commented):
Thanks for adding this! the code changes look good to me, will check the test codes later on

zhaojuanmao commented on test/distributed/fsdp/test_fsdp_comm.py:
```diff
@@ -0,0 +1,149 @@
+# Owner(s): ["oncall: distributed"]
```
nit: could we rename this test file name to be "test_fsdp_no_sync_comm" or merge with "test_fsdp_no_sync"?  as both of them are testing no_sync features

zhaojuanmao reviewed (commented):
solid and clean tests! look good from my side, just one nit. Will let @rohan-varma accept it, thanks!

zhaojuanmao commented on test/distributed/fsdp/test_fsdp_comm.py:
```diff
@@ -0,0 +1,149 @@
+        num_no_sync_iters: int,
+    ):
+        """
+        Tests the ``no_sync()`` context manager by checking the communication
```
nit: change the comment to mention that this test is generally for checking the communication cost

awgu replied:
Great catch. I see why it was confusing. I will update this.

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

**Test Plan**
- `_test_no_sync()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun4 python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

Differential Revision: [D34085750](https://our.internmc.facebook.com/intern/diff/D34085750)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

rohan-varma reviewed (approved):
Awesome work and great tests!

awgu commented on test/distributed/fsdp/test_fsdp_comm.py:
```diff
@@ -0,0 +1,148 @@
+class TestCommunication(FSDPTest):
+    """Tests ``FullyShardedDataParallel``'s collective communication usage."""
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
```
Good point. I will reduce the number of tests to 4 by always using `num_no_sync_iters = 3` and `num_sync_iters = 3`. I had both being either `1` or `3` since we originally suspected that there may be different behavior depending on how many iterations were run.

Commit:
Update on "[FSDP] Add no_sync() context manager"


**Overview**
This addresses https://github.com/pytorch/pytorch/issues/72183 and upstreams the `no_sync()` context manager for `FullyShardedDataParallel`.

**Test Plan**
- `_test_no_sync()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L66)).
- `test_communication()` is generalized from Fairscale (see [here](https://github.com/facebookresearch/fairscale/blob/89e1ae5f1631c6e038268240f92084dc2f6f2dd8/tests/nn/data_parallel/test_fsdp_grad_acc.py#L128)).

I tested for world sizes of 2 and 4 on the AWS cluster:
```
gpurun python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun4 python test/distributed/fsdp/test_fsdp_no_sync.py
gpurun python test/distributed/fsdp/test_fsdp_comm.py
gpurun4 python test/distributed/fsdp/test_fsdp_comm.py
```

Differential Revision: [D34085750](https://our.internmc.facebook.com/intern/diff/D34085750)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085750).
